### PR TITLE
menu is working

### PR DESCRIPTION
**Pull request by aeickho**
This patch added:

- Menu extensions (List, Radio button, Popup with timeout)
- spiflash write support _experimental_
- temporary small font import #55 by @maushammer 
- some new gfx functions

In Menu Configurable:

- users.csv in spi flash, lockup on dmr receive, tnx to @sellibitze
- beep tone tuning
- german date format string on main screen 

Open:

- handling of unicode char in users.csv

### CODE REVIEW

**Comment by travisgoodspeed** [comment]
I'll review this and hopefully merge it in a few hours.  Is there anything in particular where you expect trouble?

**Comment by aeickho** [comment]
No... I think the merge makes no trouble

File: applet/src/menu.c [hunk 1]
```diff
@@ -11,17 +11,28 @@
 #include "version.h"
 #include "config.h"
 #include "os.h"
+#include "spiflash.h"
+#include "addl_config.h"
 
+static wchar_t wt_addl_func[]         = L"Addl. Funct";
+static wchar_t wt_datef[]             = L"Date format";
```
Comment (aeickho) [inline applet/src/menu.c]: Is this the correct definitions for no copy to ram?

File: applet/src/addl_config.c [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  uint8_t  datef;
+  uint8_t  userscsv;
+  uint8_t  debug;
+  } global_addl_config;
+
+void init_global_addl_config_hook(void) {
+  int8_t buf[1];
+
+  spiflash_read(buf, spi_flash_addl_config_start + offset_rbeep, 1);
+  if (buf[0] == '1') {
```
Comment (aeickho) [inline applet/src/addl_config.c]: empty flash is 0xff (default setting)

File: md380-tool [hunk 3]
```diff
@@ -233,16 +292,34 @@ def hexdump(dfu,address,length=512):
...
     buf=dfu.peek(adr,length);
     i=0;
+    cbuf="";
     for b in buf:
         sys.stdout.write("%02x "%b);
         i=i+1;
-        if i%16==0: sys.stdout.write("\n");
-        elif i%8==0: sys.stdout.write(" ");
+        if(b > 32 and b < 127 ):
+            cbuf = cbuf + chr(b);
```
Comment (aeickho) [inline md380-tool]: this is an modification of hexdump output.....

File: applet/merge.py [hunk 4]
```diff
@@ -331,6 +249,62 @@ def hookbl(self,adr,handler,oldhandler=None):
...
+        0x0801b2ca, 0x0801b2f2, 0x0801b364, 0x0801ad50, 0x0801ad82, 0x0801adec, 0x0801ae1e, 0x0801245c,
+        0x080198a2, 0x0801992c];
+    ### only for debug and information addiction
+    #for adr in Create_MenuEntrylist:
+    #    merger.hookbl(adr,sapplet.getadr("create_menu_entry_hook"),0);
+
+    # additional menu hook
+    merger.hookbl(0x08012740,sapplet.getadr("create_menu_utilies_hook"),0);
+    # skip ..."unused code"
+    merger.hookstub(0x08012740+4,0x08012786+1);
```
Comment (aeickho) [inline applet/merge.py]: we own a nicer method? to skip unused code?

File: applet/src/md380-2.032.c [hunk 5]
```diff
@@ -12,8 +12,21 @@
 
 //Firmware calls to 2.032.
 
-
```
Comment (aeickho) [inline applet/src/md380-2.032.c]: what do you think about the idea to rename all firmware functions with the header md380_ to iritationen with avoiding newlib?

File: applet/src/md380-2.032.c [hunk 6]
```diff
@@ -84,3 +125,13 @@ int (*ambe_unpack)(int a1, int a2, char length, int a4)=0x08048c9d;
...
 		       int a4, short a5, short a6, int a7)=0x08051249;
+
+
+//! Functions and Variabes regarding the beep_process
+uint32_t *beep_process_unkown=0x2001d178;
+
+
+//! useful md380 firmware functions
+void (*md380_itow)(wchar_t *, int value)=0x080172ed;
+void (*md380_RTC_GetDate)(uint32_t RTC_Format, RTC_DateTypeDef *RTC_DateStruct)=0x08026461;
```
Comment (aeickho) [inline applet/src/md380-2.032.c]: example ... the function RTC_GetDate  is also in the stm lib

File: applet/src/gfx.c [hunk 7]
```diff
@@ -66,11 +84,80 @@ void lcd_background_led(int on) {
...
+void print_DebugLine_gray(void *bmp, int idx, uint64_t pos) {
+  if (global_addl_config.userscsv == 1) {
+    print_DebugLine(0x888888);
+  } else {
+    gfx_drawbmp(bmp, idx, pos);
+  }
+}
+
+
+void print_date_hook(void) {  // copy from the md380 code
```
Comment (aeickho) [inline applet/src/gfx.c]: what is your wish date format? :wink:

File: applet/src/gfx.c [hunk 8]
```diff
@@ -66,11 +84,80 @@ void lcd_background_led(int on) {
...
+      buf[ii++]='\0';
+      drawascii2(buf, 10, 50+n*10);
+      ii=0;
+      n++;
+    } else {
+      if (ii<29) buf[ii++]=DebugLine2[i];
+      }
+  }
+                                           
+  drawascii2(DebugLine1, 10,50);
```
Comment (aeickho) [inline applet/src/gfx.c]: the first line is overwritten by DebugLine1

File: applet/src/menu.c [hunk 9]
```diff
@@ -79,3 +90,325 @@ e0020000 Selected Page Index
   return main_menu(menu);
 }
 
+
+
+
+void create_menu_entry_hook(int a, void * b , void * c, void  * d, int e, int f ,int g) {
```
Comment (aeickho) [inline applet/src/menu.c]: only for debug

File: applet/src/menu.c [hunk 10]
```diff
@@ -79,3 +90,325 @@ e0020000 Selected Page Index
...
+  menu_mem->unknown_01 = 0;
+
+  create_menu_entry_hook( (*menu_id), wt_enable, menu_entry_back, menu_entry_back, 6, 2 , 1);
+  spiflash_write("1", spi_flash_addl_config_start + offset_rbeep, 1);
+  global_addl_config.rbeep = 1;
+}
+
+void create_menu_entry_rbeep_disable_screen(void) {
+  struct MENU *menu_mem;
+
```
Comment (aeickho) [inline applet/src/menu.c]: chage 0xc to sizeof(struct MENU)

File: applet/src/dmr.c [hunk 11]
```diff
@@ -48,54 +50,58 @@ void *dmr_call_end_hook(char *pkt){
...
   int src=(pkt[10]|
-      (pkt[9]<<8)|
-          (pkt[8]<<16));
+           (pkt[9]<<8)|
+           (pkt[8]<<16));
+
+  sprintf(DebugLine1, "%d -> %d", src, dst );
 
-  sprintf(DebugLine1, "%d", src );
-  sprintf(DebugLine2, "%d", dst );
```
Comment (aeickho) [inline applet/src/dmr.c]: +todo insert OSSemPend, for Debugline[12]

**Comment by phl0** [comment]
Just put @aeickho branch onto my MD380. No problems so far. Tested:

- German date format
- users.csv import

**Pull request merged**